c10/core/impl/cow/state_machine.cpp

```diff
@@ -0,0 +1,240 @@
+};
+
+cow::StateMachine::StateMachine() : state_rep_(StateId::initial) {}
+
+cow::StateMachine::~StateMachine() {
+  // If we created an impl, we are responsible for clenaing this up
+  // here.
+  Rep state_rep = state_rep_;
+  switch (state_rep.as_state_id()) {
+    case StateId::initial:
+      // Nothing to do.
+      break;
+
+    case StateId::initializing:
+      TORCH_INTERNAL_ASSERT(false); // we should not race with the destructor
+
+    case StateId::active:
+      delete &state_rep.as_state_machine_impl();
+  }
+}
+
+auto cow::StateMachine::physical_generation() -> Generation {
+  Impl* impl = maybe_get_impl();
+  return impl != nullptr ? impl->physical_generation_ : 0;
+}
+
+auto cow::StateMachine::maybe_bump(cow::ShadowStorage* maybe_shadow_storage)
+    -> void {
+  Impl* impl = maybe_get_impl();
+  if (impl == nullptr) {
+    // Any created shadow storage should be bound to the physical
+    // storage that it was created from. Hence, there should only be a
+    // shadow storage on a tensor if its own storage created it. We
+    // don't check for the specific matching of the storage and shadow
+    // storage, but we do check that the presence relationship holds.
+    //
+    // TODO: This check should probably be here, but it is actually
+    // triggering in the StaticRuntime.autogen_inner test.
+    //
+    // TORCH_INTERNAL_ASSERT(maybe_shadow_storage == nullptr);
+    return;
+  }
+
+  impl->maybe_bump(maybe_shadow_storage);
+}
+
+auto cow::StateMachine::simulate_lazy_copy(
+    cow::ShadowStorage* maybe_shadow_storage)
+    -> intrusive_ptr<cow::ShadowStorage> {
+  return ensure_initialized().simulate_lazy_copy(maybe_shadow_storage);
+}
+
+auto cow::StateMachine::maybe_get_impl() -> cow::StateMachine::Impl* {
+  return get_impl([] { return nullptr; });
+}
+
+auto cow::StateMachine::ensure_initialized() -> cow::StateMachine::Impl& {
+  return *get_impl([this]() -> c10::optional<cow::StateMachine::Impl*> {
+    Rep initial_rep(StateId::initial);
+    Rep initializing_rep(cow::StateMachine::StateId::initializing);
+    // If we fail to exchange, it means that someone else is
+    // initializing.
+    if (!state_rep_.compare_exchange_strong(initial_rep, initializing_rep)) {
+      return c10::nullopt;
+    }
+
+    auto impl = std::make_unique<cow::StateMachine::Impl>();
+    Rep new_rep(*impl);
+    TORCH_INTERNAL_ASSERT(
+        new_rep.as_state_id() == cow::StateMachine::StateId::active);
+    // We won the race to initializing: we will be the only thread
+    // that may transition to active.
+    TORCH_INTERNAL_ASSERT(
+        state_rep_.compare_exchange_strong(initializing_rep, new_rep));
+    return c10::make_optional(impl.release()); // owned by state_rep_ now
+  });
+}
+
+template <typename OnInitialFunc>
+auto cow::StateMachine::get_impl(const OnInitialFunc& on_initial_func)
+    -> StateMachine::Impl* {
+  bool seen_initializing = false;
+  do {
+    Rep state_rep = state_rep_;
```

Comment:
I'd appreciate an explicit load() here

Reply:
fair enough. done.

Reply:
If you take my suggestion about going to only one CEX, `state_rep_` now monotonically goes from nullptr to a pointer (in fact, you don't need Rep anymore, because you no longer have initializing state, and you can use nullptr as the uninitailized sigil). Because the transition is monotonic, you don't need a seq cst load. In fact, for some uses of `get_impl`, you don't need an atomic read at all, because you don't care if you race with another thread establishing the default storage impl or not. This is specifically in the mutation detection case: you are always happy to report that there is no problem, because if you're racing with a writing thread, well, you're racing with a writing thread.

Reply:
And for other uses of read, a more relaxed read is OK. I think consume is good enough.

Reply:
good point for the read before exchange. For the read in maybe_bump(), I agree that consume should do the trick since we just need to ensure that we don't see a null after a simulate is created. If they raced in a real copy-on-write world, then that means a tensor would have been having a read or write concurrent with a write to the same tensor or one its views. That's already not supported so I we don't have to be any stricter than that.

the CEX can be memory_order_acq_rel yes?

done.